File: cpp/0406根据身高重建.cpp
#include<list>
#include<vector>
#include<algorithm>

using namespace std;
static bool cmp(const vector<int> &a, const vector<int> &b)
{
    if(a[0]!=b[0])
        return a[0] > b[0];
    return a[1] < b[1];
}
vector<vector<int>> reconstructQueue(vector<vector<int>> & people)
{
    sort(people.begin(), people.end(), cmp);
    list<vector<int>> que; // list底层是链表实现，插入效率比vector高的多
    for (int i = 0; i < people.size(); i++)
    {
        int position = people[i][1]; // 插入到下标为position的位置
        std::list<vector<int>>::iterator it = que.begin();
        while (position--)
        { // 寻找在插入位置
            it++;
        }
        que.insert(it, people[i]);
    }
    return vector<vector<int>>(que.begin(), que.end());
}

File: python/0406根据身高重建.py

# 两个维度就一定要先考虑其中一个，在考虑另一个进行重排
def reconstructQueue(people: list) -> list:

    # 先按身高从大到小排序，注意身高相同时第二项大的在后面
    people=sorted(people,key=lambda x:(-x[0],x[1]))

    res=[]

    # 直接把第二项当作索引插入，从前往后遍历，符合题目要求
    for i in people:
        res.insert(i[1],i)

    return res

print(reconstructQueue([[6,0],[5,0],[4,0],[3,2],[2,2],[1,4]]))